glib/src/translate.rs

```diff
@@ -1081,56 +1174,104 @@ impl FromGlib<glib_sys::gboolean> for bool {
     }
 }
 
-impl FromGlib<u32> for char {
-    #[inline]
-    fn from_glib(val: u32) -> char {
-        char::from_u32(val).expect("Valid Unicode character expected")
-    }
-}
-
 impl FromGlib<i32> for Ordering {
     #[inline]
     fn from_glib(val: i32) -> Ordering {
         val.cmp(&0)
     }
 }
 
-impl FromGlib<u32> for Option<char> {
-    #[inline]
-    fn from_glib(val: u32) -> Option<char> {
-        match val {
-            0 => None,
-            _ => char::from_u32(val),
-        }
+/// Translate from a Glib type which can result in an undefined and/or invalid value.
+pub trait TryFromGlib<G>: Sized {
+    type Error;
+    fn try_from_glib(val: G) -> Result<Self, Self::Error>;
+}
+
+/// Error type for `TryFromGlib` when the Glib value is None.
+#[derive(Debug, PartialEq, Eq)]
+pub struct GlibNoneError;
+
+impl fmt::Display for GlibNoneError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "glib value is None")
     }
 }
 
-impl FromGlib<i32> for Option<u32> {
+impl std::error::Error for GlibNoneError {}
+
+impl<G, T: TryFromGlib<G, Error = GlibNoneError>> FromGlib<G> for Option<T> {
     #[inline]
-    fn from_glib(val: i32) -> Option<u32> {
-        if val >= 0 {
-            Some(val as u32)
-        } else {
-            None
+    fn from_glib(val: G) -> Option<T> {
+        T::try_from_glib(val).ok()
+    }
+}
+
+/// Error type for `TryFromGlib` when the Glib value can be None or invalid.
+#[derive(Debug, Eq, PartialEq)]
+pub enum GlibNoneOrInvalidError<I> {
+    Invalid(I),
```

Comment:
Maybe name this `E` instead of `I` to make it a little bit clearer that this is supposed to be an error type. Maybe even require `std::error::Error` for it

Reply:
I didn't want to force the `Error` trait on implementers initially, but on a second thought, I guess that would be better so that we can use this in interfaces with `dyn Error`.

I keep believing the name `I` is useful here to indicate that it applies to the `Invalid` variant. With the `Error` bound, don't you think this would make it clear that this is an `Error`?

Reply:
With the `Error` bound it seems fine